### DISCUSSION TITLE
module 'monai.transforms' has no attribute 'AddChannel'

### DISCUSSION THREAD

**Discussion by slefcourt27** [Project-MONAI/tutorials#1686] [Q&A] [answered]
I am using the most recent version of monai - monai_weekly-1.4.dev2413 because other versions give me an error of 

`AttributeError: 'FileFinder' object has no attribute 'find_module'
`

Trying to migrate to the most recent version and am met with this issue 

`module 'monai.transforms' has no attribute 'AddChannel'`

Is there documentation somewhere which explains which functions are deprecated and their replacements? I tried reading the official documentation but did not see this anywhere.

**Comment by KumoLiu** [comment]
Hi @slefcourt27,

> AttributeError: 'FileFinder' object has no attribute 'find_module' 

Fixed in this PR.
https://github.com/Project-MONAI/MONAI/issues/7458

> module 'monai.transforms' has no attribute 'AddChannel'

`AddChannel` has been removed since 1.3. 
https://github.com/Project-MONAI/MONAI/releases

Please check your version or perhaps reinstall it in a fresh environment.
Hope it helps, thanks.

**Reply by slefcourt27** [reply]
Hi @KumoLiu, 

One more question - What has AddChannel been replaced with?  To confirm, the AddChannel has been removed and the 
corresponding code is                 

Original
`transforms.AddChanneld(keys=["image"]),`

Updated
`transforms.EnsureChannelFirstd(keys=["image"])`

Is this correct? Thank you for your help and those previous references were helpful.

**Reply by KumoLiu** [reply, marked as answer]
Will be equal to:
`transforms.EnsureChannelFirstd(keys=["image"], channel_dim="no_channel")`
https://github.com/Project-MONAI/MONAI/blob/c0b9cc0b00459563196b9ce1d430f5d86406c5e6/monai/transforms/utility/array.py#L188

**Reply by slefcourt27** [reply]
Thank you for your response.  You have answered my question!

As an aside, I found this tutorial also https://github.com/Project-MONAI/tutorials/blob/main/modules/3d_image_transforms.ipynb and it seems that both 

`ensure_channel_first_no_channel = EnsureChannelFirstd(keys=["image", "label"], channel_dim="no_channel")`

and 

`ensure_channel_first = EnsureChannelFirstd(keys=["image", "label"])`

return the same shape of 

`torch.Size([1, 512, 512, 55])`

![Screenshot 2024-04-05 at 12 02 34 PM](https://github.com/Project-MONAI/tutorials/assets/31667320/1e66cce7-d547-477e-aaff-69801149d1f6)

**Reply by ebaldelomar** [reply]
Hi, 

I'm trying to address a similar issue and it's not clear to me how one goes about implementing this edit. I'm pretty new to Monai and monailabel. 

I’m trying to test out the case 2 example in the Quickstart documentation ([Quickstart — MONAI Label 0.8.5 Documentation](https://docs.monai.io/projects/label/en/latest/quickstart.html#use-case-2-bundle-with-customized-scripts-for-renal-substructure-segmentation)).

I got all things installed, downloaded the test data set, and started the server. Below is the error I get when trying to hit 'run' on the Auto Segmentation:

> raise ModuleNotFoundError(f"Cannot locate class or function path: ‘{__path}’.")
> ModuleNotFoundError: Cannot locate class or function path: ‘AddChanneld’.


I have searched both the monai/ and monailabel/ folders for 'AddChanneld' 


> grep --include=\*.{py} -rnw '/bmr207/nmrgrp/nmr141/.local/lib/python3.8/site-packages/' -e "AddChanneld"

Output:
> /bmr207/nmrgrp/nmr141/.local/lib/python3.8/site-packages/monailabel/tasks/infer/basic_infer.py:171:                    monai.transforms.AddChanneld(keys='image'),
> /bmr207/nmrgrp/nmr141/.local/lib/python3.8/site-packages/monailabel/tasks/infer/basic_infer.py:211:                    monai.transforms.AddChanneld(keys='pred'),
> /bmr207/nmrgrp/nmr141/.local/lib/python3.8/site-packages/tests/unit/deepedit/test_deepedit_transforms.py:16:from monai.transforms import AddChanneld
> /bmr207/nmrgrp/nmr141/.local/lib/python3.8/site-packages/tests/unit/deepedit/test_deepedit_transforms.py:184:        add_ch = AddChanneld(keys=["image", "label"])(input_data)
> /bmr207/nmrgrp/nmr141/.local/lib/python3.8/site-packages/tests/unit/deepedit/test_deepedit_transforms.py:193:        add_ch = AddChanneld(keys=["image", "label"])(input_data)
> /bmr207/nmrgrp/nmr141/.local/lib/python3.8/site-packages/tests/unit/deepedit/test_deepedit_transforms.py:202:        add_ch = AddChanneld(keys=["image", "label"])(input_data)
> /bmr207/nmrgrp/nmr141/.local/lib/python3.8/site-packages/tests/unit/deepedit/test_deepedit_transforms.py:211:        add_ch = AddChanneld(keys=["image", "label"])(input_data)
> /bmr207/nmrgrp/nmr141/.local/lib/python3.8/site-packages/tests/unit/activelearning/test_epistemic.py:24:    AddChanneld,
> /bmr207/nmrgrp/nmr141/.local/lib/python3.8/site-packages/tests/unit/activelearning/test_epistemic.py:73:                AddChanneld(keys),
> /bmr207/nmrgrp/nmr141/.local/lib/python3.8/site-packages/monai/transforms/transform.py:443:             except for example: `AddChanneld` expects (spatial_dim_1[, spatial_dim_2, ...])

The only two files that actually use 'AddChanneld' (not commented out) are: 
- tests/unit/activelearning/test_epistemic.py:73
- tests/unit/deepedit/test_deepedit_transforms.py

I found this thread but it's not clear to me how one goes about properly editing it; and if it's still the appropriate issue.

When checking version of things using:

>  python -c "import monai; monai.config.print_config()"

I get the following output if helpful:

> MONAI version: 1.3.2
> Numpy version: 1.23.4
> Pytorch version: 2.4.1+cu121
> MONAI flags: HAS_EXT = False, USE_COMPILED = False, USE_META_DICT = False
> MONAI rev id: 59a7211070538586369afd4a01eca0a7fe2e742e
> MONAI __file__: /bmr207/nmrgrp/<username>/.local/lib/python3.8/site-packages/monai/__init__.py
> 
> Optional dependencies:
> Pytorch Ignite version: 0.4.11
> ITK version: 5.4.3
> Nibabel version: 4.0.2
> scikit-image version: 0.21.0
> scipy version: 1.10.1
> Pillow version: 9.3.0
> Tensorboard version: 2.14.0
> gdown version: 5.2.0
> TorchVision version: 0.19.1+cu121
> tqdm version: 4.64.1
> lmdb version: 1.6.2
> psutil version: 7.0.0
> pandas version: 1.5.1
> einops version: 0.7.0
> transformers version: NOT INSTALLED or UNKNOWN VERSION.
> mlflow version: 2.17.2
> pynrrd version: 1.0.0
> clearml version: NOT INSTALLED or UNKNOWN VERSION.

**Reply by slefcourt27** [reply]
Hi @ebaldelomar, as @KumoLiu mentioned, the AddChannelD function has been removed since MONAI version 1.3.  I suggest that you uninstall the MONAI package and reinstall through specifying an earlier version.

**Reply by ebaldelomar** [reply]
Ah! I didn't think of that. A few questions to follow up to help my understanding if you happen to know:

1) This is likely an issue with the specific bundle/model I'm trying to implement, thus why I may need an older version?

2) When you say uninstall monai package; would that include monai and monailabel? or just monai?

Thanks so much for the response and help!

**Reply by slefcourt27** [reply]
You're welcome.

1.  It has to do that the function itself was removed from the library.  So somewhere underlying it is being called and triggering your error.  I would go line-by-line and see which is raising your ModuleNotFoundError.
2.  I would try removing and re-installing both.  If you specify the MONAI package, I think the monailabel will try to match the right version.  But you can also see which version pairs are compatible, try searching documentation or creating a new github post for that.

**Reply by RJacobArthrex** [reply]
> Thank you for your response. You have answered my question!
> 
> As an aside, I found this tutorial also https://github.com/Project-MONAI/tutorials/blob/main/modules/3d_image_transforms.ipynb and it seems that both
> 
> `ensure_channel_first_no_channel = EnsureChannelFirstd(keys=["image", "label"], channel_dim="no_channel")`
> 
> and
> 
> `ensure_channel_first = EnsureChannelFirstd(keys=["image", "label"])`
> 
> return the same shape of
> 
> `torch.Size([1, 512, 512, 55])`
> 
> ![Screenshot 2024-04-05 at 12 02 34 PM](https://private-user-images.githubusercontent.com/31667320/320065647-1e66cce7-d547-477e-aaff-69801149d1f6.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NDc3MjkzNTQsIm5iZiI6MTc0NzcyOTA1NCwicGF0aCI6Ii8zMTY2NzMyMC8zMjAwNjU2NDctMWU2NmNjZTctZDU0Ny00NzdlLWFhZmYtNjk4MDExNDlkMWY2LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTA1MjAlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUwNTIwVDA4MTczNFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTkxYTJjMmJmNzgzNjI1NTY2OTU2NTFjNzg4ZTUxMjkxMzlhMzY5OGIxOWFkMGM1MWZjNzUyNWU1ZWJlNWNjYjUmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.t6n3y_yLL1rTlRRtoEBAb_5oJmw-4qYQ4-OF_SgJc2c)

In the case provided by the tutorial, there will be no difference since a channel dimension does not exist in the input tensor. In both cases, a new channel dimension is created in axis 0 (first axis).

The difference occurs when input _**does**_ have a channel. If input dimension is `512, 512, 55, 1` with last dimension as the channel axis,
- `ensure_channel_first_no_channel` will treat all dimensions as spatial and output will be `1, 512, 512, 55, 1`.
- `ensure_channel_first` will move last channel axis to first and output `1, 512, 512, 55`.
